ports/esp32/machine_adc.c

```diff
@@ -31,178 +32,414 @@
 
 #include "driver/gpio.h"
 #include "driver/adc.h"
+#include "esp_adc_cal.h"
 
 #include "py/runtime.h"
 #include "py/mphal.h"
 #include "modmachine.h"
 
+
+#define DEFAULT_VREF 1100
+
+
+typedef struct _madcblock_obj_t {
+    mp_obj_base_t base;
+    adc_unit_t unit_id;
+    mp_int_t bits;
+    adc_bits_width_t width;
+    esp_adc_cal_characteristics_t *characteristics[ADC_ATTEN_MAX];
+} madcblock_obj_t;
+
+STATIC madcblock_obj_t madcblock_obj[] = {
+    #if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32C3
+    {{&machine_adcblock_type}, ADC_UNIT_1, 12, -1, {0}},
+    {{&machine_adcblock_type}, ADC_UNIT_2, 12, -1, {0}},
+    #elif CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
+    {{&machine_adcblock_type}, ADC_UNIT_1, 13, -1, {0}},
+    {{&machine_adcblock_type}, ADC_UNIT_2, 13, -1, {0}},
+    #endif
+};
```

Comment:
IMO the ADCBlock implementation should go in a separate file, in `machine_adcblock.c`.  Keeps it consistent with a separate class per file.

Reply:
OK. Will make this change.

Reply:
Actually, I may have spoken too soon on this one.

I've just had a quick go at splitting the ADCBlock implementation off into its own file, but this turns out to be non-trivial because of how tied together the two types are. In particular, both the `ADC()` constructor and `ADCBlock.connect()` method walk the list of ADCs to find the matching singleton object, so this array needs to be completely defined for both and its length is dependent on a bunch of platform `#ifdef`s.

I can make this work (probably with some kind of shared helper function) but it won't be as neat as the code currently is, so I thought I'd check how wedded you are to splitting them apart before I do so.

Reply:
OK, I've taken a second pass at splitting off ADCBlock with some new helper functions and I don't hate the result. I've done this as a separate commit so that the changes can be seen.